util/tracecpu/profile.go

```diff
@@ -0,0 +1,450 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracecpu
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime/pprof"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/pprof/profile"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const (
+	labelSQL        = "sql"
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+	labelOptimize   = "optimize"
+)
+
+// GlobalTopSQLCPUProfiler is the global SQL stats profiler.
+var GlobalTopSQLCPUProfiler = newTopSQLCPUProfiler()
+
+type topSQLCPUProfiler struct {
+	taskCh     chan *profileTask
+	cacheBufCh chan *profileTask
+
+	mu struct {
+		sync.Mutex
+		ept *exportProfileTask
+	}
+	collector atomic.Value
+}
+
+// newTopSQLCPUProfiler create a topSQLCPUProfiler.
+func newTopSQLCPUProfiler() *topSQLCPUProfiler {
+	return &topSQLCPUProfiler{
+		taskCh:     make(chan *profileTask, 128),
+		cacheBufCh: make(chan *profileTask, 128),
+	}
+}
+
+func (sp *topSQLCPUProfiler) Run() {
+	logutil.BgLogger().Info("cpu profiler started")
+	go sp.startCPUProfileWorker()
+	go sp.startAnalyzeProfileWorker()
+}
+
+func (sp *topSQLCPUProfiler) SetCollector(c TopSQLCollector) {
+	sp.collector.Store(c)
+}
+
+func (sp *topSQLCPUProfiler) getCollector() TopSQLCollector {
+	c, ok := sp.collector.Load().(TopSQLCollector)
+	if !ok || c == nil {
+		return nil
+	}
+	return c
+}
+
+func (sp *topSQLCPUProfiler) RegisterSQL(sqlDigest, normalizedSQL string) {
+	c := sp.getCollector()
+	if c == nil {
+		return
+	}
+	c.RegisterSQL(sqlDigest, normalizedSQL)
+}
+
+func (sp *topSQLCPUProfiler) RegisterPlan(planDigest string, normalizedPlan string) {
+	c := sp.getCollector()
+	if c == nil {
+		return
+	}
+	c.RegisterPlan(planDigest, normalizedPlan)
+}
```

Comment:
It might not be a good idea to register plan text + plan digest in the "topSQLCPUProfiler", since as a CPUProfiler it should only know (care about) digests, but not texts. This breaks the encapsulation.

Reply:
Addressed, PTAL